oci/layout/oci_dest.go

```diff
@@ -299,6 +303,90 @@ func (d *ociImageDestination) CommitWithOptions(ctx context.Context, options pri
 	return os.WriteFile(d.ref.indexPath(), indexJSON, 0644)
 }
 
+// tryHardlinkLocalFile attempts to hardlink the specified file and digest it.
+func tryHardlinkLocalFile(dest *ociImageDestination, file string) (private.UploadedBlob, bool, error) {
+	blobFile, err := os.CreateTemp(dest.ref.dir, "oci-put-blob")
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+
+	// Remove the file for os.Link() to work.
+	blobFile.Close()
+	os.Remove(blobFile.Name())
+	if err := os.Link(file, blobFile.Name()); err != nil {
+		// Fallback iff Link has failed.
+		return private.UploadedBlob{}, true, err
+	}
+
+	blobFile, err = os.Open(blobFile.Name())
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+	blobDigest, err := digest.FromReader(blobFile)
+	if err != nil {
+		blobFile.Close()
+		return private.UploadedBlob{}, false, err
+	}
+	blobPath, err := dest.ref.blobPath(blobDigest, dest.sharedBlobDir)
+	if err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+	if err := ensureParentDirectoryExists(blobPath); err != nil {
+		return private.UploadedBlob{}, false, err
+	}
+
+	// need to explicitly close the file, since a rename won't otherwise work on Windows
+	blobFile.Close()
+	if err := os.Rename(blobFile.Name(), blobPath); err != nil {
```

Comment:
We might already have a file with exactly that digest; that’s not an error.

Reply:
Will add a test for that case 👍

Reply:
It turned out that os.Rename will not fail if the file already exists but replace it.

Reply:
Oops, my mistake.